React + Snowpack Guide

melissamcewen opened the pull request:
## Changes

Adds a new React guide that will help users get started with React on Snowpack

Direct link to preview: https://snowpack-git-react-guide.pikapkg.vercel.app/guides/react

TODO:

- [x] Add images
- [x] Link to new docs about tests
- [x] Fix TOC
- [x] Set margin to 0 in minimal template
- [x] Design issues
- [x] App template minimal missing
- [x] Add certificate/social share

## Testing

TODO: 

- [ ] Add link checker test for anchors

## Docs
This is docs :) 

- [ ] Add link from main docs
- [ ] Add list of guides to main docs (now that we have more than 1)

melissamcewen commented:
Direct link to the guide in the preview [https://snowpack-git-react-guide.pikapkg.vercel.app/guides/react](https://snowpack-git-react-guide.pikapkg.vercel.app/guides/react)

FredKSchott commented:
Nice! Yea, I wonder what a good sidebar design could look like for the guides. It we assume a layout of:

```
- h1: article title
  - h2: section title
  - h2: section title
  - h2: section title
  - ...
```

Then we should redesign it for that.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+Snowpack is a great fit for [React](https://reactjs.org/) projects of any size. It's easy to get started with and can scale to projects containing thousands of components and pages without any impact on development speed. Unlike traditional React application tooling, Snowpack saves you from getting bogged down with complex bundler setups and configuration files.
+
+In this guide, you'll go from a [minimal Snowpack project](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-minimal) to a Snowpack project with React and several useful developer optimizations. In the process you'll learn:
```
"In this guide, you'll go from an empty directory to a fully configured Snowpack project with support for React and several other useful optimizations. In the process, you'll learn:"

I think the minimal template is just step one of our template, i wouldn't want devs thinking they need to set this up themselves.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+In this guide, you'll go from a [minimal Snowpack project](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-minimal) to a Snowpack project with React and several useful developer optimizations. In the process you'll learn:
+
+- How to set up a new Snowpack project
+- How to get started with React in a Snowpack project
```
~~in a Snowpack project~~ is this implied? I think safe to remove

FredKSchott replied:
Actually, I think this would be better as "Adding your first React component"

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+- How to set up a new Snowpack project
+- How to get started with React in a Snowpack project
+- Importing and using CSS and other assets inside your project
```
run-on sentence. How about "Importing CSS, Sass, Images and other assets"

(this would work if we had a mention for Sass somewhere in the CSS section. We get so many people asking about Sass, I think it makes sense to have a mention).

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+> If you want to jump to the end to see a full featured React setup, the Create Snowpack App React template comes with everything you'll learn in this tutorial plus other useful tools.
+
+Prerequisites: Snowpack is a command line tool installed from npm. This guide assumes a basic understanding of Node.js, npm, and how to run commands in the terminal. Knowledge of React is not required, Snowpack is a great way to learn React.
```
okay I know we said no exclamation marks, but mannnnnnn I would add one at the end of that last sentence if I were writing that. Maybe: "Knowledge of React is not required. (Snowpack is a great way to learn React!)"?

this may just be a bad habit of mine :)

melissamcewen replied:
We don't have to follow their standards, I'd say gonna just override here.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+- Enabling [Fast Refresh](https://reactnative.dev/docs/fast-refresh) mode for React
+- Extending Snowpack with plugins
+
+> If you want to jump to the end to see a full featured React setup, the Create Snowpack App React template comes with everything you'll learn in this tutorial plus other useful tools.
```
add link here

SirJosh3917 replied:
Perhaps a link to jump to the end of the document with a command to generate a Create Snowpack App React project?

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+In this guide, you'll go from a [minimal Snowpack project](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-minimal) to a Snowpack project with React and several useful developer optimizations. In the process you'll learn:
+
+- How to set up a new Snowpack project
+- How to get started with React in a Snowpack project
+- Importing and using CSS and other assets inside your project
+- Enabling [Fast Refresh](https://reactnative.dev/docs/fast-refresh) mode for React
+- Extending Snowpack with plugins
```
Maybe more "connecting your favorite tools to Snowpack" or "connecting your favorite dev tools"

SirJosh3917 replied:
Perhaps there could be a link at the end of each of these, linking them to their appropriate section? I found myself wanting to go to the "Fast Refresh" section but there wasn't a quick link to go to it.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+## Getting started with Snowpack and React
+
+In this step by step you'll learn how to use the [`create-snowpack-app`](https://www.snowpack.dev/#create-snowpack-app-(csa) tool to spin up a working Snowpack project and then how to add React to it.
```
I get what you were going for here, but it makes it sound like the goal is to learn how create-snowpack-app works. I'd rephrase to:

The easiest way to start a new Snowpack project is with [`create-snowpack-app`](https://www.snowpack.dev/#create-snowpack-app-(csa)). `create-snowpack-app` is a tool to quickly set up Snowpack in a new directory. For this tutorial we're going to start with the bare-bones `@snowpack/app-template-minimal` template to quickly create a new directory with some basic HTML, CSS, and JavaScript files.

Run the following commands in a new terminal window to get started:

```bash
npx create-snowpack-app my-new-project --template @snowpack/project-template-minimal
cd my-new-project
npm run start
```

SirJosh3917 replied:
> [`create-snowpack-app`](https://www.snowpack.dev/#create-snowpack-app-(csa)

unbalanced parenthesis

SirJosh3917 commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+---
```
(Don't know where to put this) when I was searching for the guide via [https://snowpack-git-react-guide.pikapkg.vercel.app/](https://snowpack-git-react-guide.pikapkg.vercel.app/), finding the guide was more difficult than I expected. Perhaps there could be a list of guides on the sidebar on the main page, which would link to the appropriate guide?

SirJosh3917 commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+In this step by step you'll learn how to use the [`create-snowpack-app`](https://www.snowpack.dev/#create-snowpack-app-(csa) tool to spin up a working Snowpack project and then how to add React to it.
+
+`create-snowpack-app` is a command line tool that creates a working Snowpack App based on a template. The template you're starting with here is `@snowpack/project-template-minimal is` a bare-bones setup where Snowpack serves an index.html, index.css, and index.js.
```
> `@snowpack/project-template-minimal is`

codetick ends later than it should

SirJosh3917 replied:
> index.html, index.css, and index.js

are all not encased in backticks whilst a few lines below, they are

> This is about as basic as it gets. When you run `npm start` Snowpack serves an `index.html`, `index.css`, and `index.js`.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+Image: screenshot of project-template-minimal, which shows "Hello world" in text on a white background
+
+This is about as basic as it gets. When you run `npm start` Snowpack serves an `index.html`, `index.css`, and `index.js`.
```
when you run commands like these, it's good to check in and confirm the expected result. 

> Lets walk through what this command just did: `TODO`

And then after that:

> Run `npm start` to see your new website up and running! You should see ... `TODO`

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+> The `README.md` contains useful information about what each file does.
+
+The first thing that you'll want to do is install React in your project. To do that, run the following command in your project directory:
```
I do like this in the same section, but I think it's still good to check in with the dev and signify "okay, lets move on to another thing". 

> Now that you have a basic project up and running, it's time to install React. To do that...

FredKSchott commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+npm install react react-dom --save
+```
+
+> In this tutorial commands use `--save` for frontend libraries and `--save-dev` for your developer tools to keep them organized separately. This is totally up to you, though.
````
Not convinced that this is required in our guide. Maybe keep it, but call it out as an "advanced" point. Maybe prefixed this section with "Advanced" and end with "This is totally up to you, though. Your project will run regardless of which you use".

SirJosh3917 commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+
+`create-snowpack-app` is a command line tool that creates a working Snowpack App based on a template. The template you're starting with here is `@snowpack/project-template-minimal is` a bare-bones setup where Snowpack serves an index.html, index.css, and index.js.
+
+```bash
+npx create-snowpack-project react-snowpack --template @snowpack/project-template-minimal
+cd react-snowpack
+npm run start
+```
````
add equivalent yarn commands, which i *believe* would be

```shell
yarn create snowpack-project react-snowpack --template @snowpack/project-template-minimal
# idk if you need to cd or what not, package doesn't exist on npm so i didn't bother checking more
```

[yarn create](https://classic.yarnpkg.com/en/docs/cli/create/)

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+## Create your first React component
+
+React uses a syntax called [JSX](https://reactjs.org/docs/introducing-jsx.html) for components. Snowpack has built in support for JSX files with the extension `.jsx`. To create your first React component in Snowpack rename `index.js` to `index.jsx`.
```
> Snowpack has built in support for JSX in files with the `.jsx` extension. That means that there's no additional plugins or configuration needed to write your first React component, but you will need to rename your `index.js` file to `index.jsx` to take advantage of this.

SirJosh3917 commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+> The `README.md` contains useful information about what each file does.
+
+The first thing that you'll want to do is install React in your project. To do that, run the following command in your project directory:
+
+To install these you can run:
```
> To do that, run the following command in your project directory:
>
> To install these you can run:

odd

melissamcewen replied:
oops haha

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+React uses a syntax called [JSX](https://reactjs.org/docs/introducing-jsx.html) for components. Snowpack has built in support for JSX files with the extension `.jsx`. To create your first React component in Snowpack rename `index.js` to `index.jsx`.
+
+> You may have used JSX in `.js` files before, but Snowpack expects JSX to be .jsx files
```
I think the comment I made above means you don't need to hit on this point twice.

SirJosh3917 commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+To install these you can run:
+
+```bash
+npm install react react-dom --save
````
equivalent yarn command maybe?

```shell
yarn add react react-dom
```

FredKSchott replied:
I think it's safe to focus on npm for now, since we use npm everywhere else in the guide. 

If we really wanted to, maybe somewhere in the intro / prerequisits section we could mention something like "If you prefer yarn or pnpm, fell free to use those as well"

SirJosh3917 commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+mv index.js index.jsx
+```
+
+> Remember that your index.html file is importing `/index.js`, not `/index.js`. Do you need to update it to match the new file? You don't because Snowpack turns `index.jsx` (which wouldn't run in the browser) into the `index.js` (which would).
````
> is importing `/index.js`, not `/index.js`

same file name twice?

SirJosh3917 replied:
> You don't because Snowpack turns `index.jsx` (which wouldn't run in the browser) into the `index.js` (which would).

add why, "which wouldn't run in the browser because it's file extension isn't .js"

FredKSchott replied:
Could be good to phrase as a question: 

> Question: Do I need to update my index.html script tag to point to "index.jsx"?
> Answer: No! Because JSX isn't supported in the browser, Snowpack builds your file to a runnable `.js` file. This is good to remember when you're referencing built files in HTML `<script src="">` and `<link href="">` elements.

I'm not too attached to the Q&A style, but teaching the larger point here is the reason it's worth calling out vs. ignoring entirely.

SirJosh3917 commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+
+Import React and add this test component `index.jsx` so you can use it.
+
+```diff
+/* Add JavaScript code here! */
+- console.log('Hello World! You did! Welcome to Snowpack :D');
++import React from 'react';
++import ReactDOM from 'react-dom';
++ReactDOM.render(<div>"HELLO REACT"</div>, document.getElementById('root'));
+```
````
using `diff` makes it more work on the user to copy paste the code, as they have to delete the pluses. should just show the entirety of the js if possible (or figure out on the preview site how to make the +s and -s not affect copying the code)

melissamcewen replied:
Hmm very true, any opinions here @drwpow @FredKSchott ?

FredKSchott replied:
As long as it's possible to select these lines for copy without the `+` and `-` attached, then I like this format.

FredKSchott replied:
I like the diff! As long as the `+` and `-` don't automatically come with each copy paste.

drwpow replied:
Diff or no, I think the indentation is off (click View File) here. Not an issue for now in draft, but would like to make sure the indents are aligned before publishing.

SirJosh3917 commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+You can now import React into `index.jsx` and render a simple test component just to make sure it's working.
+
+Import React and add this test component `index.jsx` so you can use it.
```
you've already stated your intentions in the above line

FredKSchott replied:
unnecceary line? duplicates what was said above

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+> Remember that your index.html file is importing `/index.js`, not `/index.js`. Do you need to update it to match the new file? You don't because Snowpack turns `index.jsx` (which wouldn't run in the browser) into the `index.js` (which would).
+
+You can now import React into `index.jsx` and render a simple test component just to make sure it's working.
```
nit: "import React in `index.jsx` and render`

SirJosh3917 commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
++ReactDOM.render(<div>"HELLO REACT"</div>, document.getElementById('root'));
+```
+
+If you run this code you'll get `Error: Target container is not a DOM element.` The call to `document.getElementById('root')` is failing because the HTML page has no element with that ID. Head to index.html and add one in the `<body>`
````
several minor thigns:
- `.` in backticks for error
- `index.html` not being in code ticks despite files being in code ticks everywhere else
- comma after "if you run this code" feels natural

fixed:
```
If you run this code, you'll get `Error: Target container is not a DOM element`. The call to `document.getElementById('root')` is failing because the HTML page has no element with that ID. Head to `index.html` and add one in the `<body>`.
```

SirJosh3917 commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+
+If you run this code you'll get `Error: Target container is not a DOM element.` The call to `document.getElementById('root')` is failing because the HTML page has no element with that ID. Head to index.html and add one in the `<body>`
+
+```diff
+  <body>
+-    <h1>Welcome to Snowpack!</h1>
++    <div id="root"></div>
+    <script type="module" src="/index.js"></script>
+```
````
see: previous comment related to `diff`

SirJosh3917 commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+## Organizing your files
+
+If you're going to be adding a bunch of new files, you probably don't want them in the root directory. A standard way of organizing React projects is an `src` directory for components and their associated files. Then a public directory for static assets like images, fonts, icons, and sometimes CSS. To do this you'll need to reorganize your files, then tell Snowpack about it in the configuration f ...
```
- `To do this,`
- `index.html`

FredKSchott commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
++ReactDOM.render(<div>"HELLO REACT"</div>, document.getElementById('root'));
+```
+
+If you run this code you'll get `Error: Target container is not a DOM element.` The call to `document.getElementById('root')` is failing because the HTML page has no element with that ID. Head to index.html and add one in the `<body>`
````
Whenever you give instructions that fail intentionally, it's good to make it very clear. Sometimes people will even add a comment to the code snippet itself, but in this case I don't think that's warranted since the code is fine.

Something like:

> But wait! When yuo run this code in the browser, you'll see an error: "Target container is not a DOM element." This is because the call to `document.getElementById('root')` is failing because...

A screenshot of the error popup will also highlight that this is expected

melissamcewen replied:
Hmm since the tutorial already quite long I wonder if we can just omit the error section. I'm going to try it out and see how it looks.

SirJosh3917 commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+↳ index.css
+↳ index.html
+```
+
````
maybe there should be some text here that says what the commands are doing? "Provided are some commands that give the above project structure:"

FredKSchott commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+    <script type="module" src="/index.js"></script>
+```
+
+Now it should run and you should see "HELLO REACT"
````
Section ends abruptly. Would be good to recap the main point(s?) and do some sort of celebration. You just wrote your first React component in Snowpack!!!

SirJosh3917 commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+  },
+```
+
+This tells it look at `public` instead of look at `/` for the index. For example when building `index.css`, Snowpack looks for this files in `./public/index.css`.
````
`this files`

SirJosh3917 commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+This tells it look at `public` instead of look at `/` for the index. For example when building `index.css`, Snowpack looks for this files in `./public/index.css`.
+
+And it also tells it to put the final output files in a subdirectory called `_dist_` so change that path in `index.html` from `/index.js` to `/_dist_/index.js`.
```
add a comma after "_dist_`, so change that", was surprising that the sentence turned into a command midway reading

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+Image: screenshot of the project, which shows "HELLO REACT" on a white background
+
+## Organizing your files
```
just a thought: I wonder if we want this headline to stick out to someone reading the guide but following along in their own application. That use would really care about "Customizing your project layout" or structure. They'd probably want to see the word "Configuring" somewhere in there as well, but we need to balance that with our whole "low config" vibe.

Maybe "Customizing your File Layout"? What do you think?

melissamcewen replied:
I think the word layout is a bit confusing, at least to me. I think of layout as being a front-end thing?

melissamcewen replied:
Gonna try file structure

SirJosh3917 commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+import React, {useState, useEffect} from 'react';
+
+function App() {
+  // Create the count state.
+  const [count, setCount] = useState(0);
+  // Create the counter (+1 every second).
+  useEffect(() => {
+    const timer = setTimeout(() => setCount(count + 1), 1000);
+    return () => clearTimeout(timer);
+  }, [count, setCount]);
+  // Return the App component.
```
someone reading this probably already knows react. since we're in a guide, should try to keep the amount of code minimal.

FredKSchott replied:
A stated goal here was to get you something that matched the current `create-snowpack-app` React template. If you see that template, you'll see it's basically an exact match to what you end up with here.

We also need the counter to highlight HMR & Fast Reload in the section below/

I'm personally not against the extra lines of copy-paste here, but @melissamcewen it could be worth revisiting/confirming this goal. Do we need to match the create-snowpack-app react template EXACTLY, or can we just match it in spirit? If just in spriit, then I'd agree we could probably trim some CSS, or include a Snowpack logo in the `minimal` template and then use that here instead of the React logo.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+## Organizing your files
+
+If you're going to be adding a bunch of new files, you probably don't want them in the root directory. A standard way of organizing React projects is an `src` directory for components and their associated files. Then a public directory for static assets like images, fonts, icons, and sometimes CSS. To do this you'll need to reorganize your files, then tell Snowpack about it in the configuration f ...
```
"If you're going to be adding a bunch of new files" sounds conditional, instead: 

> As your project grows...

FredKSchott replied:
Also lots of information contained in that last sentence, even I got lost a bit :) 

If it's too many steps to say in one sentence, maybe just hit the bigger point here? 

> Lets update our project to use this popular structure.

FredKSchott replied:
Also lets highlight that we don't lock you into a single cookie cutter layout: Full rewrite of this paragraph:

> If you're going to be adding a bunch of new files, you probably don't want them in the root directory. Snowpack is flexible enough to support whatever project layout that you prefer. For this guide, let's use a popular project layout from the React communtiy.
> 
> A standard way of organizing React projects is an `src` directory for components and their associated files. Then a `public` directory for static assets like images, fonts, icons, and sometimes CSS. Lets update our project to use this popular structure.

SirJosh3917 commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+  }, [count, setCount]);
+  // Return the App component.
+  return (
+    <div className="App">
+      <header className="App-header">
+        <p>
+          Edit <code>src/App.jsx</code> and save to reload.
+        </p>
+        <p>
+          Page has been open for <code>{count}</code> seconds.
+        </p>
+        <p>
+          <a
+            className="App-link"
+            href="https://reactjs.org"
+            target="_blank"
+            rel="noopener noreferrer"
+          >
+            Learn React
+          </a>
+        </p>
+      </header>
+    </div>
```
this takes up a lot of lines in a guide - can it be trimmed?

SirJosh3917 commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+
+Now include it in `index.js`
+
+```diff
+import React from 'react';
+import ReactDOM from 'react-dom';
++ import App from './App';
+
+- ReactDOM.render(<div>"HELLO WORLD"</div>, document.getElementById('root'));
++ ReactDOM.render(
++    <React.StrictMode>
++      <App />
++    </React.StrictMode>,
++    document.getElementById('root'),
++  );
+```
````
so much is being added, can probably not do the `diff` thing

FredKSchott replied:
I do like the diff, but the mixing of add/remove does make copy-paste harder. Can we move the 1 red line to above the `import App from './App';` line, so that all new lines are together?

SirJosh3917 commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+export default App;
+```
+
+Now include it in `index.js`
````
`index.jsx`

SirJosh3917 commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+```diff
+import React from 'react';
+import ReactDOM from 'react-dom';
++ import App from './App';
````
isn't it `App.jsx` or does snowpack do magic here?

melissamcewen replied:
I'm pretty sure Snowpack does magic. But @FredKSchott would know more.

FredKSchott replied:
Both are supported, but its best to get users in the habit of including the file extension

SirJosh3917 commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+> As you're doing this, you should not need to reload the page or restart Snowpack. Snowpack updates the project in the browser as you edit code, which is pretty nifty.
+
+Add this file `logo.svg` (link to file) to `src`. Now you can import it into your `App.jsx` and use it in an `img` tag to display it.
```
shouldn't assets like `logo.svg` be in `public`? earlier in the guide it said

> Then a public directory for static assets like images, fonts, icons, and sometimes CSS.

melissamcewen replied:
Good question, I'll leave this one open. I think sometimes if assets are component specific people will add them in the component directory.

FredKSchott replied:
+1 for @melissamcewen's answer here, although that's a good point of possible confusion. Maybe worth clarifying with an added line to this section like "for some static assets, you may want to locate them near the React components that use them..."

and clarifying the public line with something like "a public directory for **site** assets like images, fonts, icons, and sometimes CSS."

SirJosh3917 commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+  padding: 4px 8px;
+  border-radius: 4px;
+}
+.App p {
+  margin: 0.4rem;
+}
+.App-logo {
+  height: 40vmin;
+  pointer-events: none;
+}
+
+@media (prefers-reduced-motion: no-preference) {
+  .App-logo {
+    animation: App-logo-spin infinite 20s linear;
+  }
+}
+
+.App-header {
+  background-color: #282c34;
+  min-height: 100vh;
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+  justify-content: center;
+  font-size: calc(10px + 2vmin);
+  color: white;
+}
+
+.App-link {
+  color: #61dafb;
+}
+
+@keyframes App-logo-spin {
+  from {
+    transform: rotate(0deg);
+  }
+  to {
+    transform: rotate(360deg);
+  }
+}
```
what's the bare minimum to get the logo to spin and things to be centered okay-ishly, this is a lot of scrolling lol

SirJosh3917 commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+Image: the React logo (a blue atom) is now at the top of the page
+
+The project already has index.css for global styles, but component CSS can help keep code organized. Snowpack knows to only load this CSS when displaying the `App` component, which means it's not loading unnecessary CSS.
```
"component CSS" how about "having CSS for each component"? unless "component CSS" is an official react term, then use that cuz it's shorter

SirJosh3917 commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+You shouldn't need to restart Snowpack to see this, it should look like this:
+
+IMAGE: screenshot of the project, showing text saying "edit `src/App.js` and save to reload," text that says "Page has been open for" and the number of seconds then "seconds". A link with the text Learn React that links to the React website.
```
make sure the image says "edit `src/App.jsx` ..."

FredKSchott commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+↳ index.html
+```
+
+```bash
````
maybe add a comment: If it feels more familiar, feel free to use your computer's visual file browser as well.

SirJosh3917 commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+}
+```
+
+Image:
````
not sure if this being left blank or not is intentional so im commenting, but easily inferrable so /shrug

SirJosh3917 commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+);
++// Hot Module Replacement (HMR) - Remove this snippet to remove HMR.
++// Learn more: https://www.snowpack.dev/#hot-module-replacement
++if (import.meta.hot) {
++  import.meta.hot.accept();
++ }
```
`+ }` -> `+}` for consistency

SirJosh3917 commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+};
+```
+
+> you'll need to restart Snowpack
````
"... for the configuration changes to take effect"

SirJosh3917 commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+Install the module first by running.
+
+```bash
+npm install @snowpack/plugin-react-refresh --save-dev
````
yarn eq: `yarn add @snowpack/plugin-react-refresh --dev`

SirJosh3917 commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+- [Prettier](https://prettier.io/) – a code formatter
+
+- Tests
```
mention mocha/jest? (similar to what was done above for prettier)

FredKSchott commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+mv index.css public/index.css
+```
+
+You need to tell Snowpack about this new structure using the [Snowpack Configuration API](https://www.snowpack.dev/#configuration) in the `snowpack.config.js` file. Right now it's a bare bones scaffold that doesn't do anything. Add this to the `mount` object:
````
Good idea to mention the API, but not so fast. As a reader, I'd click it immediately and then get overwhelmed. How about:

Once your new structure is in place, you need to tell Snowpack about it. Every Snowpack project comes with a `snowpack.config.js` file for any configuration that you might need. Right now, you should see a mostly-empty, bare-bones config file. Add this to the empty `mount` object:

```
{CODE SNIPPET}
```

In the section below, you can link to the full API reference

SirJosh3917 reviewed (changes requested):
her e u go

oh forgot to mention in review but consistency with a period at the end of lines or not maybe? sometimes a line ended with one, sometimes it didnt

FredKSchott commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+  },
+```
+
+This tells it look at `public` instead of look at `/` for the index. For example when building `index.css`, Snowpack looks for this files in `./public/index.css`.
````
start with a single sentence explaining what "mount" does in general.

melissamcewen replied:
I made an attempt but LMK how it is

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+If Snowpack is still running, exit the process, you need to restart Snowpack for configuration file changes. When you start up again, if it worked, it should look the same.
+
+Create a new file at `src/App.jsx` and paste the following code into this new file to create an `App` component:
```
I'm leaning towards this `src/App.jsx` section being its own, short section. Because we outline the structure that we want to move to at the top, it feels weird to end this section with a different set of files than we started.

FredKSchott commented on www/guides/react.md:
````diff
@@ -0,0 +1,373 @@
+```diff
+import React from 'react';
+import ReactDOM from 'react-dom';
++ import App from './App';
````
lets do './App.jsx' here. We want people to get into that habit.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+When you add assets like images or CSS, Snowpack includes them in your final build. If you already know React, this process should look pretty familiar.
+
+> As you're doing this, you should not need to reload the page or restart Snowpack. Snowpack updates the project in the browser as you edit code, which is pretty nifty.
```
Snowpack automatically updates

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+> As you're doing this, you should not need to reload the page or restart Snowpack. Snowpack updates the project in the browser as you edit code, which is pretty nifty.
+
+Add this file `logo.svg` (link to file) to `src`. Now you can import it into your `App.jsx` and use it in an `img` tag to display it.
```
nit: to your `src/` directory.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+body {
+  font-family: sans-serif;
+-  margin: 2em 1em;
++  margin: 0;
```
hahaha good catch. lets make this the default in the minimal template so that we don't need this extra step.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+Image:
+
+## Adding React Fast Refresh with a Snowpack plugin
```
How about "Making Snowpack Even Faster with Fast Refresh"

or "Developing Even Faster with Fast Refresh"

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+
+## Adding React Fast Refresh with a Snowpack plugin
+
+[React Fast Refresh](https://reactnative.dev/docs/fast-refresh)? What's that? It's a developer tool that makes it much easier to see how updates to your code work in the browser. React projects are often interactive and include state. For example, this project you're building has a state that is the amount of time on the page. When developing with state it's useful not to lose it while you edit c ...
```
Running out of time, but would love to take a pass on this section later. Overall I like this new approach though.

drwpow replied:
Love the conversational tone here

FredKSchott reviewed (commented):
Woo! Overall guide structure/story looks good, so now focusing more on wordsmithing and clarity.

melissamcewen commented:
I did some refactoring of the CSS. I removed all the CSS from the base layout to make it easier to deal with. Added a new guide CSS file, as a stopgap solution until I have time to reorganize it. I noticed prism.css was for some reason making the diff so the added line was red and the removed line was green? That's an odd choice so I overrode it. I also made it so it should be easier to copy and paste from diffs, but it might be hacky? Let me know.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,361 @@
+
+Image: ideally a nice header image showing the before or after, or some illustration with elements of the Snowpack logo, React, and symbols related to speed, joy.
+
+Snowpack is a great fit for [React](https://reactjs.org/) projects of any size. It's easy to get started with and can scale to projects containing thousands of components and pages without any impact on development speed. Unlike traditional React application tooling, Snowpack saves you from getting bogged down with complex bundler setups and configuration files.
```
It's easy to get started ~~with~~ and can scale?

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,361 @@
+
+Snowpack is a great fit for [React](https://reactjs.org/) projects of any size. It's easy to get started with and can scale to projects containing thousands of components and pages without any impact on development speed. Unlike traditional React application tooling, Snowpack saves you from getting bogged down with complex bundler setups and configuration files.
+
+In this guide, you'll go from an empty directory to a fully configured Snowpack project with support for React and several other useful optimizations. In the process, you'll learn:
```
~optimizations~ features?

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,361 @@
+
+The easiest way to start a new Snowpack project is with [Create Snowpack App](<https://www.snowpack.dev/#create-snowpack-app-(csa)>). Create Snowpack App is a tool to quickly set up Snowpack in a new directory.
+
+The template you're starting with here is `@snowpack/project-template-minimal` is a bare-bones setup where Snowpack serves an `index.html`, `index.css`, and `index.js`.
```
The template that you will start with here is

or 

In this guide, you will use the `@snowpack/project-template-minimal` template to get started with a simple, bare-bones Snowpack project setup. The rest of the guide will build on this.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,361 @@
+
+The template you're starting with here is `@snowpack/project-template-minimal` is a bare-bones setup where Snowpack serves an `index.html`, `index.css`, and `index.js`.
+
+Open up your terminal and head to a directory where you want to put your new project. Now run `npx create-snowpack-project react-snowpack --template @snowpack/project-template-minimal`. This creates a new directory called `react-snowpack` and installs the `@snowpack/project-template-minimal` inside it.
```
Start with "To get started, open your terminal..."?

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,361 @@
+
+Open up your terminal and head to a directory where you want to put your new project. Now run `npx create-snowpack-project react-snowpack --template @snowpack/project-template-minimal`. This creates a new directory called `react-snowpack` and installs the `@snowpack/project-template-minimal` inside it.
+
+You can now head to that new `react-snowpack` directory with `cd react-snowpack`. And to start up Snowpack run `npm run start`. You should see your new website up and running!
```
Here and above: all commands that the user should run should look the same, stylistically. That means moving 

```
npx create-snowpack-project react-snowpack --template @snowpack/project-template-minimal
```

and 

```
cd react-snowpack
npm run start
```

into their own full code blocks. 

The text you have don't necessarily have to change,  although the create-snowpack-app command is long enough that you could probably replace that text with:

> Now run the following command in your terminal to create a new directory called `react-snowpack` that will automatically be filled with the minimal template & Snowpack setup.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,361 @@
+
+> 💡 Tip: you can also use Yarn for any of the commands in this tutorial.
+
+Now that you have a basic project up and running, it's time to install React. To install React, run the following command in your project directory:
```
install twice, maybe "it's time to add React"?

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,361 @@
+
+## Create your first React component
+
+Snowpack has built in support for JSX in files with the `.jsx` extension. That means that there's no additional plugins or configuration needed to write your first React component, but you will need to rename your `index.js` file to `index.jsx` to take advantage of this.
```
New user: "what's JSX?"

Prefix this paragraph with "React relies on a special templating language called JSX. Snowpack has..."

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,361 @@
+
+Image: screenshot of the project, which shows "HELLO REACT" on a white background
+
+You've just created your first React component in Snowpack! And also gotten your project prepared for building more with React.
```
bad grammar with "And also", i think that second sentence is unnecessary.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,361 @@
+
+You've just created your first React component in Snowpack! And also gotten your project prepared for building more with React.
+
+## Customizing your file layout
```
~~Customizing~~ Customize

Also "project layout" instead of file layout?

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,361 @@
+
+## Customizing your file layout
+
+If you're going to be adding a bunch of new files, you probably don't want them in the root directory. Snowpack is flexible enough to support whatever project layout that you prefer. For this guide, let's use a popular project layout from the React community.
```
New user: Why not?

maybe instead: "you probably don't want them crowding up your top-level root directly."

FredKSchott replied:
Something about "If you're going to be adding" is also reading funny. Maybe more "Since you'll be adding"?

FredKSchott commented on www/guides/react.md:
````diff
@@ -0,0 +1,361 @@
+mv index.css public/index.css
+```
+
+Once your new structure is in place, you need to tell Snowpack about it. Every Snowpack project comes with a `snowpack.config.js` file for any configuration that you might need. Right now, you should see a mostly-empty, bare-bones config file. Add this to the empty `mount` object:
````
"Once your new file structure" or "Once your new project structure"

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,361 @@
+  return (
+    <div className="App">
+      <header className="App-header">
++        <img src={logo} className="App-logo" alt="logo" />
```
can you add the 1 line below this one, so that people get a better idea of where this should go in their file?

FredKSchott reviewed (commented):
More edits!

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,373 @@
+date: 2020-10-01
+---
+
+Image: ideally a nice header image showing the before or after, or some illustration with elements of the Snowpack logo, React, and symbols related to speed, joy.
```
Nope, SGTM!

drwpow commented on www/guides/react.md:
````diff
@@ -0,0 +1,371 @@
+> 💡 Tip: if you want to jump to the end to see a full featured React setup, the [Create Snowpack App React template](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-react) comes with everything you'll learn in this guide plus other useful tools.
+>
+> ```bash
+> npx create-snowpack-project react-snowpack-template --template @snowpack/project-template-react
````
Is there a reason this template name is different from the snippet right below? I don’t want super beginners to be confused something isn’t working if they copied this one by mistake

drwpow replied:
Oh I’m a dummy this is the finished project. WDYT about moving this snippet to the very end, and only linking to the last heading here? That way the only snippets they see are copy-safe (but can still jump to the end if they want)

drwpow commented on www/guides/react.md:
````diff
@@ -0,0 +1,371 @@
+mv index.css public/index.css
+```
+
+Once your new project structure is in place, you need to tell Snowpack about it. Every Snowpack project comes with a `snowpack.config.js` file for any configuration that you might need. Right now, you should see a mostly empty, bare-bones configuration file. Add this to the empty `mount` object:
````
I like this explainer section, but I think as a beginner I’d be confused why `src` and why `public`. What is the purpose of each?

Also by default, Snowpack scans the entire folder; `mount` just limits the scope. So it may be more accurate now to say “we may have files we don’t want to publish online, so let’s tell Snowpack to only build specific folders.” You may also want to mention this can be a speed boost, if we’re only scanning part of the project.

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,371 @@
+  useEffect(() => {
+    const timer = setTimeout(() => setCount(count + 1), 1000);
+    return () => clearTimeout(timer);
+  }, [count, setCount]);
```
We don’t need `setCount` in the memo here (but also it’s not hurting anything)

FredKSchott replied:
some linters get mad if you don't :/ agreed it's not very important but good to do the right thing here in case anyone has one turned on by default

melissamcewen replied:
that's just copy paste from the React template so maybe change it there?

drwpow replied:
> some linters get mad if you don't :/

Yeah I always disabled those because they’re wrong! 😛  Queue another discussion about React hooks. (I’m fine for keeping, because again here it doesn’t hurt anything)

drwpow commented on www/guides/react.md:
````diff
@@ -0,0 +1,371 @@
++  );
+```
+
+> 💡 Tip: [Strict Mode](https://reactjs.org/docs/strict-mode.html) is a tool for highlighting potential problems in React code.
````
Love these little tips!

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,371 @@
+  pointer-events: none;
+}
+
+@media (prefers-reduced-motion: no-preference) {
```
❤️ ❤️ ❤️

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,371 @@
+
+@keyframes App-logo-spin {
+  from {
+    transform: rotate(0deg);
```
This affects nothing at all, but you can omit the `from [` block here as it’s inferred, to save lines

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,371 @@
+
+## Going further
+
+Great job! You're now ready to build the React project if your dreams with Snowpack.
```
❤️

drwpow replied:
(also sp: `of`0

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,371 @@
+
+Image: Certificate with share buttons for Twitter
+
+At this point you the basics and have a great starter for any React project. But if you compare with the official [Snowpack React template](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-react) you'll notice it has some other developer tools you might find useful:
```
```suggestion
At this point you have the basics and have a great starter for any React project. But if you compare with the official [Snowpack React template](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-react) you'll notice it has some other developer tools you might find useful:
```

melissamcewen replied:
you the basics lol

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,371 @@
+
+- [`@snowpack/plugin-dotenv`](https://github.com/pikapkg/snowpack/tree/master/plugins/plugin-dotenv) - Use `dotenv` in your Snowpack. This is useful for environment specific variables
+
+If you'd like to use Typescript with Snowpack and React, checkout the [Snowpack React Typescript](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-react-typescript) starter.
```
```suggestion
If you'd like to use TypeScript with Snowpack and React, check out the [Snowpack React TypeScript](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-react-typescript) starter.
```

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,371 @@
+
+If you'd like to use Typescript with Snowpack and React, checkout the [Snowpack React Typescript](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-react-typescript) starter.
+
+If you have any questions, comments, or corrections, we'd love to here from you in the Snowpack [discussion](https://github.com/pikapkg/snowpack/discussions) forum or our [Snowpack Discord community](https://discord.gg/rS8SnRk).
```
```suggestion
If you have any questions, comments, or corrections, we'd love to hear from you in the Snowpack [discussion](https://github.com/pikapkg/snowpack/discussions) forum or our [Snowpack Discord community](https://discord.gg/rS8SnRk).
```

drwpow commented:
So excited for this guide! 💯 I don’t have any major feedback; just eager to get this up

FredKSchott commented:
Exciting! @melissamcewen once you've addressed @drwpow's feedback (and updated to the latest master branch CSS) lets share this in our discord for feedback!

melissamcewen commented:
Thanks for the feedback @drwpow, @FredKSchott I think we can share it, I already shared in showcase before but that might have less traffic.

melissamcewen commented:
Added draft images. Some of the gifs might do better as videos though I think.

drwpow commented on www/guides/react.md:
````diff
@@ -0,0 +1,368 @@
+📁 public
+↳ index.css (builds to → public/index.css)
+↳ index.html (builds to → public/index.html)
+```
````
```suggestion
├── 📁 src/
│   └── index.jsx → src/index.js
└── 📁 public/
    ├── index.css → public/index.css
    └── index.html → public/index.html
```
Reading this again in the guide with everything else going on, it was hard for me to see this as a tree structure. I think using the “box drawing” unicode characters, and the convention of two dashes per file (`──`) make the file structure quicker to recognize at-a-glance

drwpow replied:
Also, I think you could just use `→` in the tree structure, and add a note below like “→ represents the file that Snowpack will build (note in some cases Snowpack will overwrite what’s there)”

drwpow commented on www/guides/react.md:
````diff
@@ -0,0 +1,368 @@
+📁 public : global assets like images, fonts, icons, and global CSS
+↳ index.css
+↳ index.html
+```
````
Also, see my below comment, but now that I’m looking at this I think a screenshot of the VS Code sidebar may be easier to recognize? That could establish familiarity with what developers look at

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,368 @@
+- console.log('Hello World! You did! Welcome to Snowpack :D');
++import React from 'react';
++import ReactDOM from 'react-dom';
++ReactDOM.render(<div>"HELLO REACT"</div>, document.getElementById('root'));
```
```suggestion
- console.log('Hello World! You did! Welcome to Snowpack :D');
+ import React from 'react';
+ import ReactDOM from 'react-dom';
+ ReactDOM.render(<div>"HELLO REACT"</div>, 
  document.getElementById('root'));
```

drwpow commented on www/guides/react.md:
````diff
@@ -0,0 +1,368 @@
+```diff
+  <body>
+-    <h1>Welcome to Snowpack!</h1>
++    <div id="root"></div>
````
```suggestion
  <body>
-   <h1>Welcome to Snowpack!</h1>
+   <div id="root"></div>
+   <script type="module" src="/index.js"></script>
  </body>
```

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,368 @@
+    <h1>Welcome to Snowpack!</h1>
+    <div id="root"></div>
+-    <script type="module" src="/index.js"></>
++    <script type="module" src="/_dist_/index.js"></script>
```
```suggestion
  <body>
    <h1>Welcome to Snowpack!</h1>
    <div id="root"></div>
-   <script type="module" src="/index.js"></script>
+   <script type="module" src="/_dist_/index.js"></script>
  </body>
```

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,368 @@
++      <App />
++    </React.StrictMode>,
++    document.getElementById('root'),
++  );
```
```suggestion
  import React from 'react';
  import ReactDOM from 'react-dom';
+ import App from './App.jsx';
- ReactDOM.render(<div>"HELLO WORLD"</div>, document.getElementById('root'));
+ ReactDOM.render(
+   <React.StrictMode>
+     <App />
+   </React.StrictMode>,
+   document.getElementById('root'),
+ );
```

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,368 @@
+    <div className="App">
+      <header className="App-header">
++        <img src={logo} className="App-logo" alt="logo" />
+        <p>
```
```suggestion
  function App() {
    // Create the count state.
    const [count, setCount] = useState(0);
    // Create the counter (+1 every second).
    useEffect(() => {
      const timer = setTimeout(() => setCount(count + 1), 1000);
      return () => clearTimeout(timer);
    }, [count, setCount]);
    // Return the App component.
    return (
      <div className="App">
        <header className="App-header">
+       <img src={logo} className="App-logo" alt="logo" />
        <p>
```

drwpow commented on www/guides/react.md:
````diff
@@ -0,0 +1,368 @@
+
+```diff
+import logo from './logo.svg';
++import './App.css';
````
```suggestion
  import logo from './logo.svg';
+ import './App.css';
```

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,368 @@
++// Learn more: https://www.snowpack.dev/#hot-module-replacement
++if (import.meta.hot) {
++  import.meta.hot.accept();
++}
```
```suggestion
  ReactDOM.render(
    <React.StrictMode>
      <App />
    </React.StrictMode>,
    document.getElementById('root'),
  );
+ // Hot Module Replacement (HMR) - Remove this snippet to remove HMR.
+ // Learn more: https://www.snowpack.dev/#hot-module-replacement
+ if (import.meta.hot) {
+   import.meta.hot.accept();
+ }
```

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,368 @@
+  plugins: [
++    '@snowpack/plugin-react-refresh'
+  ],
+};
```
```suggestion
  module.exports = {
    mount: {
      public: '/',
      src: '/_dist_',
    },
-   plugins: []
+   plugins: ['@snowpack/plugin-react-refresh'],
  };
```

drwpow commented on www/guides/react.md:
```diff
@@ -0,0 +1,424 @@
+
+The project already has index.css for global styles. For CSS that's only for a specific component, a common design pattern is to add it in a CSS file with the same base name as the component. The style file for `App.jsx` would be `App.css` with this pattern.
+
+> 💡 Tip: Snowpack has built-in support for [CSS Modules](https://www.snowpack.dev/#import-css-modules) and if you'd like to use SASS there is an official [SASS Plugin](https://www.snowpack.dev/#sass).
```
```suggestion
> 💡 Tip: Snowpack has built-in support for [CSS Modules](https://www.snowpack.dev/#import-css-modules) and if you'd like to use Sass there is an official [Sass Plugin](https://www.snowpack.dev/#sass).
```

melissamcewen commented:
Changed the base

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,350 @@
+
+[React Fast Refresh](https://reactnative.dev/docs/fast-refresh)? What's that? It's a developer tool that makes it much easier to see how updates to your code work in the browser. React projects are often interactive and include state. For example, this project you're building has a state that is the amount of time on the page. When developing with state it's useful not to lose it while you edit c ...
+
+Start by enabling [Hot Module Replacement](https://www.snowpack.dev/#hot-module-replacement), a requirement for Fast Refresh. Snowpack automatically refreshes to show changes in your code. Hot Module Replacement only updates the components that changed without refreshing the whole page. You can enable it for React with a small snippet of code in `index.jsx`.
```
clarify the difference: 

> Start by enabling [Hot Module Replacement](https://www.snowpack.dev/#hot-module-replacement) in your project. HMR is the system that lets Snowpack push updates to the browser without a full page refresh,  a requirement for Fast Refresh. You can enable HMR for React by adding a small snippet of code to your `src/index.jsx` file.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,350 @@
+
+Start by enabling [Hot Module Replacement](https://www.snowpack.dev/#hot-module-replacement), a requirement for Fast Refresh. Snowpack automatically refreshes to show changes in your code. Hot Module Replacement only updates the components that changed without refreshing the whole page. You can enable it for React with a small snippet of code in `index.jsx`.
+
+Head to `index.jsx` and add this snippet at the bottom:
```
Don't think this line is needed with the new edit on the paragraph above (see comment above)

FredKSchott commented on www/guides/react.md:
````diff
@@ -0,0 +1,350 @@
++ }
+```
+
+Now when you change `App.jsx` the changes show without the whole page refreshing.
````
"Now when you change `App.jsx` the page updates to show your changes without a full refresh."

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,350 @@
+
+<div class="frame"><img src="/img/guides/react/hmr.gif" alt="GIF showing code side by side with the app. A change in made to App.jsx and it shows immediately when the file is changed. The counter keeps counting uninterrupted." class="screenshot"/></div>
+
+The counter on the page still resets to 0, even though the page itself hasn't fully refreshed. This can be annoying if you're trying to debug state. React Fast Refresh solves this.
```
"HMR can save you time on its own, but you may notice in the example above that the counter on the page still resets to 0. This can slow down your development, especially when you're trying to debug a specific component state problem. Lets enable Fast Refresh to preserve component state across updates.

To enable Fast Refresh, we'll need to install the "@snowpack/plugin-react-refresh" package. This package is a Snowpack plugin, which you can use to enhance or customize Snowpack with all sorts of new behaviors. To start, install the package in your project:

FredKSchott replied:
(this comment is can replace line 305-309)

FredKSchott commented on www/guides/react.md:
````diff
@@ -0,0 +1,350 @@
+npm install @snowpack/plugin-react-refresh --save-dev
+```
+
+To enable it head to `snowpack-config.js` and add it to the array of plugins
````
"Once installed, you'll need to add the plugin to your Snowpack configuration file so that Snowpack knows to use it".

FredKSchott commented on www/guides/react.md:
````diff
@@ -0,0 +1,350 @@
+  };
+```
+
+> 💡 Tip: you'll need to restart Snowpack for the configuration changes to take effect
````
instead of having this in a blockquote, start the next paragraph with this instruction. Blockquotes are only used above as additional info and not important/required steps.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,350 @@
+
+> 💡 Tip: you'll need to restart Snowpack for the configuration changes to take effect
+
+Now make an edit to `App.jsx`. If this worked you'll see it without the counter stopping or losing its value.
```
"Restart Snowpack to apply the new plugin, and then try changing the `App.jsx`  component again. If Fast Refresh is working properly you ill see the counter keep its value across changes, without resetting to zero.

FredKSchott commented on www/guides/react.md:
```diff
@@ -0,0 +1,350 @@
+
+## Going further
+
+Great job! You're now ready to build the React project of your dreams with Snowpack.
```
Good to call out that this is the end. "Congratulations! You've reached the end of our guide, and are now ready to..."

FredKSchott reviewed (approved):
Fantastic! I left a bunch of comments, but if all are addressed then I'm +1 on merging!!! If you have any questions lets chat today to resolve, I'd love to see this merged by EOD

melissamcewen commented on www/guides/react.md:
````diff
@@ -0,0 +1,350 @@
+  },
+```
+
+<img src="/img/guides/react/folderstructure.jpg" alt="The original file configuration had Snowpack building the directory structure the same as the directories in the project, including root. Now the config builds only src and public. Src to the dist folder and public to root." />
````
yeah this is much better!
